File: php/tests/AWSSignatureTest.php
<?php

namespace Tests\Auth;

use Auth\AWSSignature;

class AWSSignatureTest extends TestUtils
{
    private $auth;

    protected function setUp(): void
    {
        parent::setUp();
        $this->auth = new AWSSignature('test-access-key', 'test-secret-key', 'us-east-1', 's3');
    }

    public function testValidAWSSignature(): void
    {
        $request = $this->createTestRequest('GET', [
            'Authorization' => 'AWS4-HMAC-SHA256 ' .
                'Credential=test-access-key/20240101/us-east-1/s3/aws4_request, ' .
                'SignedHeaders=host;x-amz-date, ' .
                'Signature=test-signature',
            'X-Amz-Date' => '20240101T000000Z'
        ]);
        $this->runAuthTest($this->auth, $request, 200, false);
    }

    public function testInvalidAccessKey(): void
    {
        $request = $this->createTestRequest('GET', [
            'Authorization' => 'AWS4-HMAC-SHA256 ' .
                'Credential=wrong-access-key/20240101/us-east-1/s3/aws4_request, ' .
                'SignedHeaders=host;x-amz-date, ' .
                'Signature=test-signature',
            'X-Amz-Date' => '20240101T000000Z'
        ]);
        $this->runAuthTest($this->auth, $request, 401, true);
    }

    public function testMissingAuthorizationHeader(): void
    {
        $request = $this->createTestRequest('GET', [
            'X-Amz-Date' => '20240101T000000Z'
        ]);
        $this->runAuthTest($this->auth, $request, 401, true);
    }

    public function testMissingXAmzDateHeader(): void
    {
        $request = $this->createTestRequest('GET', [
            'Authorization' => 'AWS4-HMAC-SHA256 ' .
                'Credential=test-access-key/20240101/us-east-1/s3/aws4_request, ' .
                'SignedHeaders=host;x-amz-date, ' .
                'Signature=test-signature'
        ]);
        $this->runAuthTest($this->auth, $request, 401, true);
    }

    public function testMalformedAuthorizationHeader(): void
    {
        $request = $this->createTestRequest('GET', [
            'Authorization' => 'AWS4-HMAC-SHA256',
            'X-Amz-Date' => '20240101T000000Z'
        ]);
        $this->runAuthTest($this->auth, $request, 401, true);
    }

    public function testGetType(): void
    {
        $this->assertAuthType($this->auth, 'aws');
    }
} 

File: go/src/auth/aws_signature_test.go
package auth

import (
	"net/http"
	"testing"
)

func TestAWSSignature(t *testing.T) {
	// Create a new AWSSignature instance with test credentials
	auth := NewAWSSignature("test-access-key", "test-secret-key", "us-east-1")

	// Test cases
	tests := []TestCase{
		{
			Name: "Valid AWS signature",
			Request: CreateTestRequest("GET", map[string]string{
				"Authorization": "AWS4-HMAC-SHA256 " +
					"Credential=test-access-key/20240101/us-east-1/s3/aws4_request, " +
					"SignedHeaders=host;x-amz-date, " +
					"Signature=test-signature",
				"X-Amz-Date": "20240101T000000Z",
			}),
			ExpectedStatus: http.StatusOK,
			ExpectedError:  false,
		},
		{
			Name: "Invalid access key",
			Request: CreateTestRequest("GET", map[string]string{
				"Authorization": "AWS4-HMAC-SHA256 " +
					"Credential=wrong-access-key/20240101/us-east-1/s3/aws4_request, " +
					"SignedHeaders=host;x-amz-date, " +
					"Signature=test-signature",
				"X-Amz-Date": "20240101T000000Z",
			}),
			ExpectedStatus: http.StatusUnauthorized,
			ExpectedError:  true,
		},
		{
			Name: "Missing Authorization header",
			Request: CreateTestRequest("GET", map[string]string{
				"X-Amz-Date": "20240101T000000Z",
			}),
			ExpectedStatus: http.StatusUnauthorized,
			ExpectedError:  true,
		},
		{
			Name: "Missing X-Amz-Date header",
			Request: CreateTestRequest("GET", map[string]string{
				"Authorization": "AWS4-HMAC-SHA256 " +
					"Credential=test-access-key/20240101/us-east-1/s3/aws4_request, " +
					"SignedHeaders=host;x-amz-date, " +
					"Signature=test-signature",
			}),
			ExpectedStatus: http.StatusUnauthorized,
			ExpectedError:  true,
		},
		{
			Name: "Malformed Authorization header",
			Request: CreateTestRequest("GET", map[string]string{
				"Authorization": "AWS4-HMAC-SHA256",
				"X-Amz-Date":    "20240101T000000Z",
			}),
			ExpectedStatus: http.StatusUnauthorized,
			ExpectedError:  true,
		},
	}

	for _, tc := range tests {
		t.Run(tc.Name, func(t *testing.T) {
			RunAuthTest(t, auth, tc)
		})
	}
}

func TestAWSSignatureGetType(t *testing.T) {
	auth := NewAWSSignature("access-key", "secret-key", "region")
	if auth.GetType() != "aws" {
		t.Errorf("GetType() = %v, want %v", auth.GetType(), "aws")
	}
} 